doc/dh_virtualenv.1.rst
=============
dh-virtualenv
=============

------------------------------------------------------
deploy a Python package in a self-contained virtualenv
------------------------------------------------------

:Author: Jyrki Pulliainen / Spotify AB <jyrki@spotify.com>
:Copyright: Copyright (C) 2013, Spotify AB. Licensed under
    the GNU General Public License version 2 or later
:Manual section: 1
:Manual group: DebHelper

SYNOPSIS
========

**dh_virtualenv** [*OPTIONS*]


DESCRIPTION
===========

``dh-virtualenv`` is a tool that aims to combine Debian packaging with
self-contained virtualenv based Python deployments. To do this, the
package extends debhelper's sequence by providing a new command in
sequence, ``dh_virtualenv``, which effectively replaces following
commands from the sequence:

 * ``dh_auto_install``
 * ``dh_python2``
 * ``dh_pycentral``
 * ``dh_pysupport``

In the sequence the ``dh_virtualenv`` is inserted right after
``dh_perl``.

OPTIONS
=======

-p PACKAGE, --package=PACKAGE		Act on the package named PACKAGE
-N PACKAGE, --no-package=PACKAGE	Do not act on the specified PACKAGE
-v, --verbose				Turn on verbose mode.
--extra-index-url			Pass extra index URL to pip
--preinstall=PACKAGE			Preinstall a PACKAGE before running pip.
--pip-tool=PIP_TOOL			Tool used to install requirements.
--extra-pip-arg				Extra arg for the pip executable.
--extra-virtualenv-arg			Extra arg for the virtualenv executable.
--index-url				Base URL for PyPI server.
--setuptools				Use setuptools instead of distribute.
--install-suffix=SUFFIX			Override virtualenv installation suffix
--upgrade-pip				Force upgrade pip in virtualenv
--requirements=FILE			Use FILE for requirements
--setuptools-test			Run `setup.py test` upon build.
--python=PATH				Use Python interpreter at PATH
--builtin-venv				Use built-in venv of Python 3
--skip-install				Don't run ``pip install .``

QUICK GUIDE FOR MAINTAINERS
===========================

1. Build depend on `python` or `python-all` and `dh-virtualenv`
2. Add `${python:Depends}` to Depends
3. Add `python-virtualenv` to dh's `--with` option

SEE ALSO
========

Online documentation can be found at
https://dh-virtualenv.readthedocs.io/en/latest.

This package should also ship with the complete documentation under
`/usr/share/doc/dh-virtualenv`.


doc/changes.rst
===========
 Changelog
===========

Following list contains the most notable changes by version.
For a full list, consult the `git history`_ of the project.

.. _`git history`: https://github.com/spotify/dh-virtualenv/commits/master


1.2 (UNRELEASED)
================

* New option :option:`--upgrade-pip-to` for increased build stability (#266) [@jhermann]


1.1
===

* Support new style shebangs generated by recent pip (#226) [@nailor]
* Add :option:`--extras` option (#243) [@jhermann]
* Python 3.4 and 3.5 added to test environments (#238) [@jhermann]
* New build dependendcies (dh-exec + python-sphinx-rtd-theme) (#231) [@labeneator]
* Disallow building a package whilst within an activated virtualenv (#224) [@lamby]
* Use ``python -m pip`` instead of direct pip calls (#219) [@moritz]
* Ignore :option:`--extra-pip-arg` in call for :option:`--upgrade-pip` (#197) [@jhermann]
* buildsystem: Allow to specify a virtualenv name (#180) [@dzen]
* docs: Improved structure, new chapters [@jhermann]
* docs: Fix reference to pbuilder's USENETWORK option (#246) [@mkohler]
* Fix setuptools and pip setup when using built-in virtualenv with
  `--system-site-packages` (#247) [@lucasrangit]


1.0
===

* **Backwards incompatible** Change the default install root to
  ``/opt/venvs``. This is due to the old installation root
  (``/usr/share/python``) clashing with Debian provided Python
  utilities. To maintain the old install location, use
  :envvar:`DH_VIRTUALENV_INSTALL_ROOT` and point it to
  ``/usr/share/python``.
* **Backwards incompatible** By default, do not run `setup.py test`
  upon building. The :option:`--no-test` flag has no longer has any
  effect. To get the old behaviour, use the
  :option:`--setuptools-test` flag instead.
* **Backwards incompatible** Buildsystem: Move files into build folder
  in install step instead of build step. Thanks to `Ludwig Hähne
  <https://github.com/Pankrat>`_ for the patch!
* Deprecate :option:`--pypi-url` in favour of :option:`--index-url`
* Support upgrading pip to the latest release with :option:`--upgrade-pip`
  flag.
* Buildsystem: Add support for :envvar:`DH_UPGRADE_PIP`,
  :envvar:`DH_UPGRADE_SETUPTOOLS` and :envvar:`DH_UPGRADE_WHEEL`. Thanks
  to `Kris Kvilekval <https://github.com/kkvilekval>`_ for the
  implementation!
* Buildsystem: Add support for custom requirements file location
  using :envvar:`DH_REQUIREMENTS_FILE` and for custom ``pip`` command
  line arguments using :envvar:`DH_PIP_EXTRA_ARGS`. Thanks to `Einar
  Forselv <https://github.com/einarf>`_ for implementing!
* Fixing shebangs now supports multiple interpreters. Thanks `Javier
  Santacruz <https://github.com/jvrsantacruz>`_!
* Allow a custom pip executable via :option:`--pip-tool` flag. Thanks
  `Anthony Sottile <https://github.com/asottile>`_ for the
  implementation!
* Fix handling of shebang lines for cases where interpreter was
  wrapped in quotes. Thanks to `Kamil Niechajewicz
  <https://github.com/noizex>`_ for fixing!
* Support extra arguments to be passed at virtualenv using
  :option:`--extra-virtualenv-arg`. Thanks to `Julien Duponchelle
  <https://github.com/noplay>`_ for the fix.

0.11
====

* Allow passing explicit filename for `requirements.txt` using
  :option:`--requirements` option. Thanks to `Eric Larson
  <https://github.com/ionrock>`_ for implementing!
* Ensure that venv is configured before starting any daemons. Thanks
  to `Chris Lamb <https://github.com/lamby>`_ for fixing this!
* Make sure `fix_activate_path` updates all activate scripts. Thanks
  to `walrusVision <https://github.com/walrusVision>`_ for fixing
  this!

0.10
====

* **Backwards incompatible** Fix installation using the built-in
  virtual environment on 3.4. This might break installation on Python
  versions prior to 3.4 when using :option:`--builtin-venv` flag.
  Thanks to `Elonen <https://github.com/elonen>`_ for fixing!
* Honor :envvar:`DH_VIRTUALENV_INSTALL_ROOT` in build system. Thanks to
  `Ludwig Hähne <https://github.com/Pankrat>`_ for implementing!
* Allow overriding virtualenv arguments by using the
  :envvar:`DH_VIRTUALENV_ARGUMENTS` environment variable when using the
  build system. Thanks to `Ludwig Hähne <https://github.com/Pankrat>`_
  for implementing!
* Add option to skip installation of the actual project. In other
  words using :option:`--skip-install` installs only the dependencies
  of the project found in requirements.txt. Thanks to `Phillip
  O'Donnell <https://github.com/phillipod>`_ for implementing!
* Support custom installation suffix instead of the package name via
  :option:`--install-suffix`. Thanks to `Phillip O'Donnell
  <https://github.com/phillipod>`_ for implementing!

0.9
===

* Support using system packages via a command line flag
  :option:`--use-system-packages`. Thanks to `Wes Mason
  <https://github.com/1stvamp>`_ for implementing this feature!
* Introduce a new, experimental, more modular build system. See the
  :doc:`usage` for documentation.
* Respect the :envvar:`DEB_NO_CHECK` environment variable.

0.8
===

* Support for running triggers upon host interpreter update. This new
  feature makes it possible to upgrade the host Python interpreter
  and avoid breakage of all the virtualenvs installed with
  dh-virtualenv. For usage, see the the :doc:`tutorial`. Huge thanks to
  `Jürgen Hermann <https://github.com/jhermann>`_ for implementing
  this long wanted feature!
* Add support for the built-in ``venv`` module. Thanks to `Petri
  Lehtinen <https://github.com/akheron>`_!
* Allow custom ``pip`` flags to be passed via the
  :option:`--extra-pip-arg` flag. Thanks to `@labeneator
  <https://github.com/labeneator>`_ for the feature.

0.7
===

* **Backwards incompatible** Support running tests. This change
  breaks builds that use distutils. For those cases a flag
  :option:`--no-test` needs to be passed.
* Add tutorial to documentation
* Don't crash on debbuild parameters ``-i`` and ``-a``
* Support custom source directory (debhelper's flag ``-D``)

0.6
===


dh_virtualenv.py
# -*- coding: utf-8 -*-
# Copyright (c) 2013 Spotify AB

# This file is part of dh-virtualenv.

# dh-virtualenv is free software: you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation, either version 2 of the
# License, or (at your option) any later version.

# dh-virtualenv is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with dh-virtualenv. If not, see
# <http://www.gnu.org/licenses/>.

import os
import re
import shutil
import subprocess
import tempfile

ROOT_ENV_KEY = 'DH_VIRTUALENV_INSTALL_ROOT'
DEFAULT_INSTALL_DIR = '/usr/share/python/'


class Deployment(object):
    def __init__(self, package, extra_urls=None, preinstall=None, pypi_url=None,
                 setuptools=False, python=None, sourcedirectory=None, verbose=False):
        self.package = package
        install_root = os.environ.get(ROOT_ENV_KEY, DEFAULT_INSTALL_DIR)
        self.virtualenv_install_dir = os.path.join(install_root, self.package)
        self.debian_root = os.path.join(
            'debian', package, install_root.lstrip('/'))
        self.package_dir = os.path.join(self.debian_root, package)
        self.bin_dir = os.path.join(self.package_dir, 'bin')

        self.extra_urls = extra_urls if extra_urls is not None else []
        self.preinstall = preinstall if preinstall is not None else []
        self.pypi_url = pypi_url
        self.log_file = tempfile.NamedTemporaryFile()
        self.verbose = verbose
        self.setuptools = setuptools
        self.python = python
        self.sourcedirectory = '.' if sourcedirectory is None else sourcedirectory

    def clean(self):
        shutil.rmtree(self.debian_root)

    def create_virtualenv(self):
        virtualenv = ['virtualenv', '--no-site-packages']

        if self.setuptools:
            virtualenv.append('--setuptools')

        if self.verbose:
            virtualenv.append('--verbose')

        if self.python:
            virtualenv.extend(('--python', self.python))

        virtualenv.append(self.package_dir)
        subprocess.check_call(virtualenv)

        # We need to prefix the pip run with the location of python
        # executable. Otherwise it would just blow up due to too long
        # shebang-line.
        self.pip_prefix = [
            os.path.join(self.bin_dir, 'python'),
            os.path.join(self.bin_dir, 'pip'),
        ]
        if self.verbose:
            self.pip_prefix.append('-v')

        self.pip_prefix.append('install')

        if self.pypi_url:
            self.pip_prefix.append('--pypi-url={0}'.format(self.pypi_url))
        self.pip_prefix.extend([
            '--extra-index-url={0}'.format(url) for url in self.extra_urls
        ])
        self.pip_prefix.append('--log={0}'.format(self.log_file.name))

    def pip(self, *args):
        return self.pip_prefix + list(args)

    def install_dependencies(self):
        # Install preinstall stage packages. This is handy if you need
        # a custom package to install dependencies (think something
        # along lines of setuptools), but that does not get installed
        # by default virtualenv.
        if self.preinstall:
            subprocess.check_call(self.pip(*self.preinstall))

        requirements_path = os.path.join(self.sourcedirectory, 'requirements.txt')
        if os.path.exists(requirements_path):
            subprocess.check_call(self.pip('-r', requirements_path))

    def fix_shebangs(self):
        """Translate /usr/bin/python and /usr/bin/env python sheband
        lines to point to our virtualenv python.
        """
        grep_proc = subprocess.Popen(
            ['grep', '-l', '-r', '-e', r'^#!.*bin/\(env \)\?python',
             self.bin_dir],
            stdout=subprocess.PIPE
        )
        files, stderr = grep_proc.communicate()
        files = files.strip()
        if not files:
            return

        pythonpath = os.path.join(self.virtualenv_install_dir, 'bin/python')
        for f in files.split('\n'):
            subprocess.check_call(
                ['sed', '-i', r's|^#!.*bin/\(env \)\?python|#!{0}|'.format(
                    pythonpath),
                 f])

    def fix_activate_path(self):
        """Replace the `VIRTUAL_ENV` path in bin/activate to reflect the
        post-install path of the virtualenv.
        """
        virtualenv_path = 'VIRTUAL_ENV="{0}"'.format(
            self.virtualenv_install_dir)
        pattern = re.compile(r'^VIRTUAL_ENV=.*$', flags=re.M)

        with open(os.path.join(self.bin_dir, 'activate'), 'r+') as fh:
            content = pattern.sub(virtualenv_path, fh.read())
            fh.seek(0)
            fh.truncate()
            fh.write(content)

    def install_package(self):
        setup_path = os.path.join(self.sourcedirectory)
        subprocess.check_call(self.pip(setup_path))
